### ISSUE TITLE
Experiment Settings for Section 5.4 Learning Auxiliary Labels

### ISSUE THREAD

**Issue by VincentPisztora** [AvivNavon/AuxiLearn#4]
Hi there,

Could you post the code for the experiments in section 5.4 (Learning Auxiliary Labels) of the paper?

Thanks!

**Comment by AvivNavon** [comment]
Hi,

The code for learning auxiliary targets is available here: https://github.com/AvivNavon/AuxiLearn/tree/master/experiments/oxford_pet#auxilearn-for-generating-auxiliary-task

**Comment by VincentPisztora** [comment]
Thanks for the quick response!

I'd like to replicate the results in Table 3 of your paper so I'm looking for the the "hyperparameter" settings for each of the datasets.
In particular, could you provide the best settings you found for each dataset of the following:

- optimizers for the main and auxiliary networks (e.g. [SGD,momentum=0.9,nesterov=True], [Adam,beta_1=0.9,beta_2=0.999])
- initial learning rate and schedules for each optimizer (main and auxiliary) (e.g. [1e-1,Cosine(T_max,eta_min,last_epoch)], [1e-4,Constant])
- batch size (default value in code is 16)
- epochs (default value in code is 150)
- auxiliary loss scale constant (default value in code is .25)
- weight decay values for the main and auxiliary networks (default in code is 5e-3)
- "auxgrad-every" value (default in code is 30)
- the size of the auxiliary set (default in code is 0.0084)
- the auxiliary network architecture. The paper notes that "the architectures of the auxiliary and primary networks 
were set the same" - does that mean that in the CIFAR experiments for example the auxiliary network is a separate VGG16?
- any data augmentations used during training (I didn't see any for the Pet data but perhaps the usual shift and flip is used for the Cifars?)
- loss functions (I see in the code focal loss with focusing constant = 2 for the Pet dataset)

Appreciate the guidance!

**Comment by IdanAchituve** [comment]
Hi,
We don't have the exact hyper-parameters, but we specified what we used and searched over in the Appendix for all datasets. You can assume that if something is not specified it is according to the default value in the code.
Regarding the auxiliary architecture network in CIFAR - both are VGG-16.
Regarding data augmentations - We used random crop and horizontal flip. In the fine-grained datasets we first resized the image (can also be seen in the code: [link](https://github.com/AvivNavon/AuxiLearn/blob/2c32f5cb548714ad3efe5c804003a30d6f012e2b/experiments/oxford_pet/data.py#L113)).
Regarding the loss function - we used the focal loss with constant = 2 in all experiments.

Cheers,
Idan

**Comment by VincentPisztora** [comment]
Hi all,

Thanks for the info.

Could you clarify whether you used the flip and shift augmentations for CIFAR in the Learning Auxiliary Labels experiments in particular?
Using the augmentations I'm able to obtain ~60.5% test accuracy on the 5% CIFAR10 using STL and the following settings: 
- VGG16
- batch_size: 16
- weight_decay: 5e-3
- optimizer: sgd,nesterov momentum = 0.9
- lr schedule: cosineDecay(1e-3)
- epochs: 200 (at 150 test accuracy is ~58.5%)

Without augmentation and using the same STL settings as above I find a test accuracy ~50% at 150 epochs and ~50% again at 200 (approx. the reported performance).

One other question - do you use the entropy penalty used by the MAXL authors when training the "sub-classification" auxiliary task using Auxilearn?
Appendix B.3 indicates that MAXL was tuned across {.2,.4,.6} entropy weights but I didn't see reference to this term for Auxilearn tuning.
I also looked for it in the Pet code but didn't spot anything. 

Thanks!

**Comment by IdanAchituve** [comment]
For CIFAR-10/100 and SVHN we used a batch size 0f 128. Indeed it is not an optimal batch size and the performance can be improved with a smaller one. Thanks for noticing and pointing that out. We used the following augmentations:
```
normalize = transforms.Normalize(
                        mean=[0.4914, 0.4822, 0.4465],
                        std=[0.2023, 0.1994, 0.2010],
                    )
train_transform = transforms.Compose([
                transforms.RandomCrop(32, padding=4),
                transforms.RandomHorizontalFlip(),
                transforms.ToTensor(),
                normalize,
            ])
```

In AuxiLearn we didn't need an entropy penalty, since the method by design prevents degenerate solutions (which is the reason for this penalty in MAXL)

**Comment by VincentPisztora** [comment]
Hi all,

Apologies for the delayed follow-up.
FYI I wasn't able to replicate with the warm restart cosine decay either.
Since I am using my own codebase it could very well be an issue outside these parameter settings.
If I am able to replicate eventually I will post another note here.

Thanks for the detailed answers!
